### DISCUSSION TITLE
[Bug] java.lang.ClassCastException: TUGIAssumingTransport can not be cast to EnhancedInstance when use trift to connect to hive with kerberos

### DISCUSSION THREAD

**Discussion by ZERONE-GO** [apache/skywalking#12580] [Q&A]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

When use trift to connect to Hive with kerberos open, it will use TUGIAssumingTransport instead. Currently trift plugin only enhanced the "org.apache.thrift.transport.TSocket" and "org.apache.thrift.transport.TNonblockingSocket", so it will throw the java.lang.ClassCastException as follow:

![tup](https://github.com/user-attachments/assets/67c6599d-d09a-422f-bc69-052da575eada)

### What you expected to happen

You can check ERROR under skywalking-api.log. The traceId is leaked when skywalking report trace segment to skywalking server. 

### How to reproduce

Use thrift with open kerberos to connect to Hive. 

### Are you willing to submit a pull request to fix on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
This is a kind of error, this scenario is not supported. No code is instrumenting your related callings.
The plugins are contributed by the community, so, you need to debug and understand the current thrift plugin, and try to fix it.

**Reply by ZERONE-GO** [reply]
The problem happened in follow code under "org.apache.skywalking.apm.plugin.thrift.client.TServiceClientInterceptor#onConstruct"

> Object dynamicField = ((EnhancedInstance) protocol.getTransport()).getSkyWalkingDynamicField();

The code convert the protocol.getTransport() to EnhancedInstance without checking the transport has been enhanced or not.

**Reply by wu-sheng** [reply]
We don't do that. The check is usually hiding the missing part. And it costs the extra payload to other cases.

**Reply by ZERONE-GO** [reply]
Sorry, I did not get the point.

In my investigation, TServiceClientInterceptor used to enhance the `org.apache.thrift.TServiceClient`, and in the consutor of `TServiceClient`, it will inject the implementation class of `TProtocol` interface.  Now the trift plugin only enhance the subclass of "org.apache.thrift.transport.TSocket" and "org.apache.thrift.transport.TNonblockingSocket".

When we use other `TProtocol` like `TUGIAssumingTransport `(which is provide by hadoop) to connect to hive, then the skywalking agent can not convert the TProtocol to enhanced class, then the problem  happen, that is why I think it should be a bug of trift plugin. 

The follow code is I think how to fix the problem.

            TTransport transport = protocol.getTransport();
            // check the transport has been enhanced or not
            if (transport instanceof EnhancedInstance) {
                Object dynamicField = ((EnhancedInstance) transport).getSkyWalkingDynamicField();
                objInst.setSkyWalkingDynamicField(Objects.isNull(dynamicField) ? "UNKNOWN" : dynamicField);
            } else {
                objInst.setSkyWalkingDynamicField("UNKNOWN");
            }

**Reply by wu-sheng** [reply]
Do you understand why it propagates something and add info into dynamic field? If not, how would you expect just skip as UNKNOWN is good?

**Reply by ZERONE-GO** [reply]
Of course skip as UNKNOWN is not good, the better way should be add the interceptor for other TProtocol  subclasses, however, the key  is we can not cover all those cases, then we need some backup solution for those cases.

**Reply by wu-sheng** [reply]
The only right move is to add new codes to cover unsupported case, rather than adding this kind of checking to impact normal case.

That is the right engineering way.

**Reply by wu-sheng** [reply]
Error message is a good thing. It is a tip to help you find out why.

**Reply by ZERONE-GO** [reply]
For this case,  I think the problem is caused by the code of trift plugin is not robust enough. According to my investigation, at least 13 types of subclass provided by hadoop jar library, but skywalking trift plugin only consider the implementation under org.apache.thrift:libthrift:0.13.0. And it convert to EnhansedInstance without any check. 

Not only the Error message is output in skywalking log, it will stop the agent  exist the span, and then some other wired problems happened in your skywalking trace log. ex, "more than xxx spans have been created", "duplicated traceid for different request."

**Reply by wu-sheng** [reply]
I don't know what we are discussing here. If you want to criticize codes, go ahead. We are very clear why we don't put type check.
This is a project consensus.
